File: client-android/serenada-core/src/test/java/app/serenada/core/fakes/FakeSignalingTransport.kt
package app.serenada.core.fakes

import app.serenada.core.call.SignalingClient
import app.serenada.core.call.SignalingMessage
import app.serenada.core.call.SignalingTransport

internal class FakeSignalingTransport(
    override val kind: SignalingClient.TransportKind,
) : SignalingTransport {

    var connectCalls = 0
        private set
    var closeCalls = 0
        private set
    val sentMessages = mutableListOf<SignalingMessage>()

    private var onOpen: (() -> Unit)? = null
    private var onMessage: ((SignalingMessage) -> Unit)? = null
    private var onClosed: ((String) -> Unit)? = null

    override fun connect(
        host: String,
        onOpen: () -> Unit,
        onMessage: (SignalingMessage) -> Unit,
        onClosed: (String) -> Unit,
    ) {
        connectCalls++
        this.onOpen = onOpen
        this.onMessage = onMessage
        this.onClosed = onClosed
    }

    override fun send(message: SignalingMessage) {
        sentMessages += message
    }

    override fun close() {
        closeCalls++
    }

    // ---- Test-side drivers ----

    fun simulateOpen() {
        onOpen?.invoke()
    }

    fun simulateClose(reason: String = "transport-closed") {
        onClosed?.invoke(reason)
    }

    fun simulateMessage(message: SignalingMessage) {
        onMessage?.invoke(message)
    }
}

File: client-ios/SerenadaCore/Tests/SerenadaCoreTests/Fakes/FakeSignalingTransport.swift
import Foundation
@testable import SerenadaCore

@MainActor
final class FakeSignalingTransport: SignalingTransport {
    let kind: TransportKind

    private(set) var connectCalls = 0
    private(set) var closeCalls = 0
    private(set) var sentMessages: [SignalingMessage] = []

    private var onOpen: (() -> Void)?
    private var onMessage: ((SignalingMessage) -> Void)?
    private var onClosed: ((String) -> Void)?

    init(kind: TransportKind) {
        self.kind = kind
    }

    func connect(
        host: String,
        onOpen: @escaping () -> Void,
        onMessage: @escaping (SignalingMessage) -> Void,
        onClosed: @escaping (String) -> Void
    ) {
        connectCalls += 1
        self.onOpen = onOpen
        self.onMessage = onMessage
        self.onClosed = onClosed
    }

    func send(_ message: SignalingMessage) {
        sentMessages.append(message)
    }

    func close() {
        closeCalls += 1
    }

    // MARK: - Test Drivers

    func simulateOpen() {
        onOpen?()
    }

    func simulateClose(_ reason: String = "transport-closed") {
        onClosed?(reason)
    }

    func simulateMessage(_ message: SignalingMessage) {
        onMessage?(message)
    }

    func clearSentMessages() {
        sentMessages.removeAll()
    }
}
